add pseudocode for protocol flow to appendix

ttaubert opened the pull request:
Alternative to #30, moving the whole pseudocode protocol flow description to the appendix.

Should we reference this somewhere?

cc @chris-wood

chris-wood commented on draft-bar-cfrg-spake2plus.md:
```diff
@@ -483,6 +483,70 @@ Thanks to Ben Kaduk and Watson Ladd, from which this specification originally em
+
+## Protocol Run
+
+A full SPAKE2+ protocol run initiated by the prover will look as follows:
```
```suggestion
A full SPAKE2+ protocol run initiated by the prover will look as follows,
where Transmit and Receive are shorthand for sending and receiving
a message to the peer:
```

chris-wood commented on draft-bar-cfrg-spake2plus.md:
```diff
@@ -483,6 +483,70 @@ Thanks to Ben Kaduk and Watson Ladd, from which this specification originally em
+
+   return (Z, V)
+~~~
+
```
```suggestion
## Transcript Computation

def ComputeTranscript(Context, idProver, idVerifier, M, N, shareP, shareV, Z, V, w0):
   TT = len(Context) || Context
      || len(idProver) || idProver
      || len(idVerifier) || idVerifier
      || len(M) || M
      || len(N) || N
      || len(shareP) || shareP
      || len(shareV) || shareV
      || len(Z) || Z
      || len(V) || V
      || len(w0) || w0

## Key Schedule Computation

def ComputeKeySchedule(TT):
   K_main = Hash(TT)
   K_confirmP || K_confirmV = KDF(nil, K_main, "ConfirmationKeys")
   K_shared = KDF(nil, K_main, "SharedKey")
   return K_confirmP, K_confirmV, K_shared
```

chris-wood commented on draft-bar-cfrg-spake2plus.md:
```diff
@@ -483,6 +483,70 @@ Thanks to Ben Kaduk and Watson Ladd, from which this specification originally em
+A full SPAKE2+ protocol run initiated by the prover will look as follows:
+
+~~~
+Prover:
```
```suggestion
Prover(Context, idProver, idVerifier, w0):
```

chris-wood commented on draft-bar-cfrg-spake2plus.md:
```diff
@@ -483,6 +483,70 @@ Thanks to Ben Kaduk and Watson Ladd, from which this specification originally em
+   (x, X) = ProverInit()
+   Transmit(X)
+   Y = Receive()
+   (Z, V) = ProverFinish(x, Y)
```
```suggestion
   (Z, V) = ProverFinish(x, Y)
   TT = ComputeTranscript(Context, idProver, idVerifier, M, N, X, Y, Z, V, w0)
   K_confirmP, K_confirmV, K_shared = ComputeKeySchedule(TT)
   expected_confirmV = MAC(K_confirmV, X)
   confirmV = Receive()
   if not_equal_constant_time(expected_confirmV, confirmV):
      raise "invalid confirmation message"
   
   confirmP = MAC(K_confirmP, Y)
   Transmit(confirmP)
```

chris-wood commented on draft-bar-cfrg-spake2plus.md:
```diff
@@ -483,6 +483,70 @@ Thanks to Ben Kaduk and Watson Ladd, from which this specification originally em
+Verifier:
+   X = Receive()
+   (Y, Z, V) = VerifierFinish(X)
+   Transmit(Y)
```
```suggestion
   Transmit(Y)
   TT = ComputeTranscript(Context, idProver, idVerifier, M, N, X, Y, Z, V, w0)
   K_confirmP, K_confirmV, K_shared = ComputeKeySchedule(TT)
   confirmV = MAC(K_confirmV, X)
   Transmit(confirmV)
   
   expected_confirmP = MAC(K_confirmP, Y)
   confirmP = Receive()
   if not_equal_constant_time(expected_confirmP, confirmP):
      raise "invalid confirmation message"
   
   confirmP = MAC(K_confirmP, Y)
   Transmit(confirmP)
```

chris-wood reviewed (commented):
I left some suggestions that fill out the entire protocol, including transcript computation and confirmation. It's a lot more words, and I'm not sure if it's helpful, but if we're going to given pseudocode it seems useful to be complete?

chris-wood commented on draft-bar-cfrg-spake2plus.md:
```diff
@@ -483,6 +483,70 @@ Thanks to Ben Kaduk and Watson Ladd, from which this specification originally em
+   Y = Receive()
+   (Z, V) = ProverFinish(x, Y)
+
+Verifier:
```
```suggestion
Verifier(Context, idProver, idVerifier, w0):
```

ttaubert commented:
I was considering too whether to include the whole flow... seems like a good idea! I've cleaned it up a bit, added a few parameters and removed the ones defined by the ciphersuite.

chris-wood commented on draft-bar-cfrg-spake2plus.md:
```diff
@@ -483,6 +483,117 @@ Thanks to Ben Kaduk and Watson Ladd, from which this specification originally em
+P, p, and h are defined by the chosen ciphersuite.
+
+## Prover
+
```
```suggestion

The Prover's behavior consists of two functions, ProverInit and ProverFinish, which
are described below.

```

chris-wood commented on draft-bar-cfrg-spake2plus.md:
```diff
@@ -483,6 +483,117 @@ Thanks to Ben Kaduk and Watson Ladd, from which this specification originally em
+   return (Y, Z, V)
+~~~
+
+## Verifier
```
```suggestion
## Verifier

The Verifier's behavior consists of a single function, VerifierFinish, which
is described below.
```

chris-wood commented on draft-bar-cfrg-spake2plus.md:
```diff
@@ -483,6 +483,117 @@ Thanks to Ben Kaduk and Watson Ladd, from which this specification originally em
+   return (Z, V)
+~~~
+
+## Transcript Computation
```
```suggestion
## Transcript Computation

Both Prover and Verifier share the same function to compute the protocol
transcript, ComputeTranscript, which is described below.
```

chris-wood commented on draft-bar-cfrg-spake2plus.md:
```diff
@@ -483,6 +483,117 @@ Thanks to Ben Kaduk and Watson Ladd, from which this specification originally em
+~~~
+
+## Key Schedule Computation
+
```
```suggestion
Both Prover and Verifier share the same function to compute
the key schedule, ComputeKeySchedule, which is described below.

```
